### PR TITLE
Dev-python/mkdocs-material: material theme for mkdocs [please reassign]

### PR DESCRIPTION
**Pull request by AndrewAmmerlaan**
The popular material theme for mkdocs and deps.

This has to all go in one PR, because of all the cross dependencies.
Some dependencies of mkdocs-material depend on mkdocs-material to compile their documentation, these circular dependencies are unavoidable but can easily be circumvented by emerging with USE="-doc" first.

See also: https://github.com/gentoo/guru/issues/3#issuecomment-612142126
And: https://github.com/gentoo/gentoo/pull/15286

### CODE REVIEW

**Comment by AndrewAmmerlaan** [comment]
> Any new eclasses need to be mailed and discussed in gentoo-dev mailing list.

Alright, I'll write an email about this then.

> I am not a fan of what's done here, but you may get positive feedback from other devs. It'd maybe be even more favored if you could integrate using doxygen to it.
Building docs isn't / shouldn't be a complex task.

I just added support for doxygen. Building docs isn't that complex, but it does take quite a few lines to do it correctly with mkdocs:

```
...

IUSE="doc"

....

BDEPEND="doc? ( 
           $(python_gen_any_dep "dev-python/mkdocs[${PYTHON_USEDEP}] other- deps"  ) 
)"

....

python_compile_all() {
	default
	if use doc; then
		mkdocs build || die "Failed to make docs"
		# Colliding files found by ecompress:
		rm site/sitemap.xml.gz || die
		HTML_DOCS=( "site/." )
	fi
}
```

And since this is the same every time, I figured it might as well be an eclass, instead of doing the same over and over again.

> (If mkdocs gains popularity it'd be easier to justify a need for new eclass, but now it seems like it's only used by handful of ebuilds and not worth it...)

As I argued here: https://github.com/gentoo/gentoo/pull/15232 the absence of mkdocs-material makes that 39% of the packages that *could* compile docs with mkdocs cannot because of the missing dependency. And statistically, I would expect that for every 14 packages that currently use distutils_enable_sphinx, there would be 1 package that has documentation building with mkdocs. 

It might just be that I've been dealing with a disproportionate amount of ebuilds using mkdocs, but I definitely feel like such an eclass might be useful.

**Comment by AndrewAmmerlaan** [comment]
> > Any new eclasses need to be mailed and discussed in gentoo-dev mailing list.
> Alright, I'll write an email about this then.

@juippis I tried to write an email to the mailing list like you suggested, but either I did it wrong, or I just don't have access. Should I send my email to you instead, so you can forward it to the mailing list?

**Comment by juippis** [comment]
You need to be subscribed to gentoo-dev in order to send mail there. You can unsubscribe later after the discussion if you wish.

**Comment by AndrewAmmerlaan** [comment]
> You need to be subscribed to gentoo-dev in order to send mail there. You can unsubscribe later after the discussion if you wish.

Aah so that is where it went wrong, I sent the email, and subscribed afterwards. I've tried again, and it seems to have worked now, Thanks

**Comment by AndrewAmmerlaan** [comment]
> Any new eclasses need to be mailed and discussed in gentoo-dev mailing list.

Considering that there haven't been any further replies to my email, does that mean the eclass is fine?

**Comment by juippis** [comment]
Usually yes, although I'd like @mgorny to approve it.

**Comment by AndrewAmmerlaan** [comment]
Version bumped mkdocs, and added py3_8 to it and deps in: https://github.com/gentoo/gentoo/pull/16366

Will, add py3_8 to these packages as well, after that PR is merged.

**Comment by juippis** [comment]
Hi, can you once again tell me what is needed before this should be reviewed? Does #16366 need to be merged first?

**Comment by AndrewAmmerlaan** [comment]
> Hi, can you once again tell me what is needed before this should be reviewed? Does #16366 need to be merged first?

Hi,
Yes, if #16366 is merged first, I can add python3.8 to the packages here afterwards.

**Comment by AndrewAmmerlaan** [comment]
> Hi, can you once again tell me what is needed before this should be reviewed? Does #16366 need to be merged first?

@juippis I have updated and rebased this PR and it is now ready for merge. I know it is a big one, but there is no way to split this up because these packages depend on mkdocs-material for building docs, and mkdocs-material depends on these packages. It would really be awesome if we could get this merged asap, because the way mkdocs works makes it a huge headache to maintain part of it in ::gentoo and part of it in ::guru. Mkdocs builds with the latest python available, which is py3.8 now, but the packages in ::guru still only have py3.7 which makes mkdocs fail to find the plugins as they haven't been installed with the latest python version that mkdocs is available for. Therefore all ::guru packages that depend on mkdocs are currently broken. I will also merge these commits to ::guru which will fix it, but in the future it would be awesome to be able to update all mkdocs pakcages to new python targets at once to avoid this problem, therefore I _really_ would like to move these packages to ::gentoo.

File: dev-python/bracex/metadata.xml
Comment (juippis) [inline dev-python/bracex/metadata.xml:8]: Can we have @Alessandro-Barbieri's ack on being added as maintainer for multiple packages here?

**Review by juippis — APPROVED** [review]
Please don't fix the issues I've reviewed, or make 2 separate commits and **don't** squash. I don't want to go through them all again.

The circular deps are are not user-friendly and I believe some more elegant solution can be come up with, but that can be dealt after this's been merged. 

This is good imho, just need @Alessandro-Barbieri's ack. I can edit metadata.xml files after a couple of days if there's no answer.

File: dev-python/mkdocs-git-revision-date-localized-plugin/mkdocs-git-revision-date-localized-plugin-0.8.ebuild
Comment (AndrewAmmerlaan) [reply dev-python/mkdocs-git-revision-date-localized-plugin/mkdocs-git-revision-date-localized-plugin-0.8.ebuild:31]: Hmm, strange, I'm not getting this error, perhaps my configuration is different and it somehow does succeed in auto-detecting an email address. In any case it should probably be easily fixable by adding:

```
python_prepare_all() {
	# mkdocs-git-revision-date-localized-plugin's tests need git repo
	if use test; then
		git init
		git config --global user.email "you@example.com" || die
		git config --global user.name "Your Name" || die
		git add .
		git commit -m 'init'
	fi

	distutils-r1_python_prepare_all
}
```

(But I'll do that later and leave this PR as is for now)

**Comment by Alessandro-Barbieri** [comment]
Yes, add me as co-maintainer

**Comment by AndrewAmmerlaan** [comment]
Thank you @juippis !

**Pull request closed without merging**
